### PR TITLE
refactor: decouple foreign functions from WasmBase.

### PR DESCRIPTION
**Pull request by mathetake**
It seems unnecessary to have foreign functions managed by WasmBase because it's just using global lookup tables. 

Signed-off-by: Takeshi Yoneda <takeshi@tetrate.io>

### CODE REVIEW

**Review by PiotrSikora — CHANGES_REQUESTED** [review]
I don't think it works right now, but foreign functions must be coupled with WasmBase in order to enforce capability restrictions.

**Comment by mathetake** [comment]
No I don't think so. The decoupled codes are only used in the "allowed" case - inside of exports::call_foreign_function which is replaced with the stub when capability not given: https://github.com/proxy-wasm/proxy-wasm-cpp-host/blob/master/src/wasm.cc#L109.

**Review by PiotrSikora — APPROVED** [review]
> No I don't think so. The decoupled codes are only used in the "allowed" case - inside of exports::call_foreign_function which is replaced with the stub when capability not given: https://github.com/proxy-wasm/proxy-wasm-cpp-host/blob/master/src/wasm.cc#L109.

Right, I guess there is some context missing. The end goal is to be able to allow/deny specific arguments, e.g. you should be able to allow only `http_call(cluster="httpbin")` or `foreign_function(name="foo", args="bar")` and deny everything else. The list of allowed arguments would be scoped to a given `WasmBase`, which is why the current code might work better.

Having said that, it's not implemented yet, so feel free to merge this if you need it.

File: src/exports.cc [hunk 1]
```diff
@@ -36,6 +36,24 @@ ContextBase *contextOrEffectiveContext() {
 // of current_context_.
 extern thread_local uint32_t effective_context_id_;
 
+// Lookup table of foreign functions, using a pointer to avoid the initialization fiasco.
+std::unordered_map<std::string, WasmForeignFunction> *foreign_functions = nullptr;
+
+WasmForeignFunction getForeignFunction(std::string_view function_name) {
+  auto it = foreign_functions->find(std::string(function_name));
```
Comment (PiotrSikora) [inline src/exports.cc:43]: I think you need to add guard here, otherwise you might be dereferencing `nullptr`.

**Review by PiotrSikora — CHANGES_REQUESTED** [review]

File: src/exports.cc (hunk 1 above)
Comment (mathetake) [reply src/exports.cc]: https://github.com/proxy-wasm/proxy-wasm-cpp-host/pull/190/commits/7893fc75855a0057741b71917224860186037d45 done!

**Comment by mathetake** [comment]
> The end goal is to be able to allow/deny specific arguments,
 
Oh I see. I think then we have to refactor around all the exports functions, not only this foreign function calls in order to couple with WasmBase. Thanks for letting me know the context. 

The initial intent of this was I just wanted to reduce the complexity in the wasm.cc/wasm.h. I will go ahead and merge since I believe this could make them a bit less complex and readable (plus added doc!)🙂

**Review by PiotrSikora — APPROVED** [review]

**Pull request merged**